File: 815. Bus Routes.java
class Solution {
    public int numBusesToDestination(int[][] routes, int source, int target) {
        HashMap<Integer, ArrayList<Integer>> adj = new HashMap<>();
        for(int i = 0; i < routes.length; i++){
            int[] route = routes[i];
            for(int stop : route){
                ArrayList<Integer> temp = adj.getOrDefault(stop, new ArrayList<>());
                temp.add(i);
                adj.put(stop, temp);
            }
        }

        if(source == target) return 0;
        if(!adj.containsKey(source) || !adj.containsKey(target)) return -1;

        int count = 0;
        Queue<Integer> q = new LinkedList<>();
        boolean[] visited = new boolean[routes.length];
        for(int route : adj.get(source)){
            visited[route] = true;
            for(int stops : routes[route]){
                q.offer(stops);
            }
        }

        while(!q.isEmpty()){
            int size = q.size();
            for(int i = 0; i < size; i++){
                int stop = q.poll();
                if(stop == target) return count + 1;

                for(int route : adj.get(stop)){
                    if(visited[route]) continue;
                    for(int j = 0; j < routes[route].length; j++){
                        int nextStop = routes[route][j];
                        q.offer(nextStop);
                    }
                    visited[route] = true;
                }
            }
            count++;
        }

        return -1;
    }
}

File: 815. Bus Routes.cpp
class Solution {
public:
    int numBusesToDestination(std::vector<std::vector<int>>& routes, int source, int target) {
        if (source == target) return 0;
        std::unordered_map<int, std::vector<int>> adj;

        for (int i = 0; i < (int)routes.size(); i++) {
            for (int stop : routes[i]) {
                adj[stop].push_back(i);
            }
        }

        if (adj.find(source) == adj.end()) return -1;
        std::unordered_set<int> visited;
        std::queue<int> q;
        q.push(source);
        for (int route : adj[source]) {
            for (int stop : routes[route]) {
                if (stop != source) q.push(stop);
            }
            visited.insert(route);
        }

        int ans = 1;
        while (!q.empty()) {
            int size = (int)q.size();
            while (size != 0) {
                int station = q.front();
                q.pop();
                if (station == target) return ans;

                for (int route : adj[station]) {
                    if (visited.count(route)) continue;
                    for (int stop : routes[route]) {
                        if (stop != station) q.push(stop);
                    }
                    visited.insert(route);
                }

                size--;
            }
            ans++;
        }

        return -1;
    }
};